sys/uri_parser/uri_parser.c

```diff
@@ -0,0 +1,178 @@
+
+    /* cannot have empty scheme */
+    if (uri[0] == ':') {
+        return NULL;
+    }
+
+    /* first character should be ALPHA */
+    if (!(((uri[0] >= 'A') && (uri[0] <= 'Z')) ||
+          ((uri[0] >= 'a') && (uri[0] <= 'z')))) {
+        NULL;
+    }
+
+    char *p = strchr(uri, ':');
+
+    /* '//' should follow ':' */
+    if (!((p[1] != '\0') && (p[2] != '\0') &&
+          (p[1] == '/')  && (p[2] == '/'))) {
+        return NULL;
+    }
+
+    result->scheme = uri;
+    result->scheme_len = p - uri;
+
+    /* skip '://' */
+    return p + 3;
+}
+
+static char *_consume_authority(uri_parser_result_t *result, char *uri)
+{
+    assert(uri);
+
+    /* search until first '/' */
+    char *authority_end = strchr(uri, '/');
+
+    result->host = uri;
+    result->host_len = authority_end - uri;
+
+    /* check for userinfo within authority */
+    char *userinfo_end = strchr(uri, '@');
+
+    /* check if match is within authority */
+    if (userinfo_end && (userinfo_end < authority_end)) {
+        result->userinfo = uri;
+        result->userinfo_len = userinfo_end - uri;
+        /* shift host part beyond userinfo and '@' */
+        result->host += result->userinfo_len + 1;
+        result->host_len -= result->userinfo_len + 1;
+    }
+
+    /* check for port after host part */
+    /* repeat until last ':' because of multiple occurrences in IPv6 addresses */
+    char *port_begin, *p = result->host;
+    while ((p < authority_end) && (p != NULL)) {
+        port_begin = p;
+        p = strchr(p + 1, ':');
+    }
+
+    /* check if match is within authority */
+    if (port_begin && (port_begin < authority_end)) {
+        result->port = port_begin + 1;
+        result->port_len = authority_end - result->port;
+        /* cut host part before port and ':' */
+        result->host_len -= result->port_len + 1;
+    }
+
+    /* this includes the '/' */
+    return authority_end;
+}
+
+static char *_consume_path(uri_parser_result_t *result, char *uri)
+{
+    assert(uri);
+    assert(uri[0] == '/');
+
+    /* uri already points to beginning of path, including '/' */
+    result->path = uri;
+    result->path_len = strlen(uri);
+
+    /* check for query start '?' */
+    char *path_end = strchr(uri, '?');
+
+    if (path_end) {
+        result->query = path_end + 1;
+        /* do not count '?' */
+        result->query_len = result->path_len - (path_end - uri) - 1;
+        /* cut path part before query and '?' */
+        result->path_len -= result->query_len + 1;
+    }
+
+    return (result->query + result->query_len);
+}
+
+static int _parse_absolute(uri_parser_result_t *result, char *uri)
+{
+    if (NULL == (uri = _consume_scheme(result, uri))) {
+        return -1;
+    }
+    if (NULL == (uri = _consume_authority(result, uri))) {
+        return -1;
+    }
+
+    /* we expect '\0', i.e., end of string */
+    uri = _consume_path(result, uri);
+    if (uri[0] != '\0') {
+        return -1;
+    }
+
+    return 0;
```

Comment:
This looks similar to the `_parse_relative` function.
```suggestion
    return _parse_relative(result, uri);
```